File: test/performance/data-structures/stack/stack.test.cpp
/**
 * Stack Benchmark Suite
 * Measures performance of stack operations (LIFO)
 * Compiled to benchmark executable by CMake
 */

#include <benchmark/benchmark.h>
#include <stack>

const int MILLION = 1'000'000;

// Test 1: 1M push
static void Push_1M(benchmark::State& state) {
  for (auto _ : state) {
    std::stack<int> stack;
    for (int i = 0; i < MILLION; ++i) {
      stack.push(i);
    }
    benchmark::DoNotOptimize(stack);
  }
}
BENCHMARK(Push_1M)->Name("1M push");

// Test 2: 1M push & pop
static void Push_Pop_1M(benchmark::State& state) {
  for (auto _ : state) {
    std::stack<int> stack;
    for (int i = 0; i < MILLION; ++i) {
      stack.push(i);
    }
    for (int i = 0; i < MILLION; ++i) {
      stack.pop();
    }
    benchmark::DoNotOptimize(stack);
  }
}
BENCHMARK(Push_Pop_1M)->Name("1M push & pop");

// Benchmark entry point
BENCHMARK_MAIN();

File: test/performance/data-structures/stack/stack.test.mjs
import { Stack } from '../../../../dist/esm/index.mjs';
import { Stack as SdslStack } from 'js-sdsl';
import Benchmark from 'benchmark';
import { magnitude } from '../../../utils/index.mjs';

const suite = new Benchmark.Suite();

const { MILLION } = magnitude;

// Test 1: 1M push
suite.add('1M push', function() {
  const stack = new Stack();
  for (let i = 0; i < MILLION; i++) {
    stack.push(i);
  }
  this.val = stack;
});

suite.add('1M push (js-sdsl)', function() {
  const stack = new SdslStack();
  for (let i = 0; i < MILLION; i++) {
    stack.push(i);
  }
  this.val = stack;
});

suite.add('Native JS Array 1M push', function() {
  const arr = [];
  for (let i = 0; i < MILLION; i++) {
    arr.push(i);
  }
  this.val = arr;
});

// Test 2: 1M push & pop
suite.add('1M push & pop', function() {
  const stack = new Stack();
  for (let i = 0; i < MILLION; i++) {
    stack.push(i);
  }
  for (let i = 0; i < MILLION; i++) {
    stack.pop();
  }
  this.val = stack;
});

suite.add('1M push & pop (js-sdsl)', function() {
  const stack = new SdslStack();
  for (let i = 0; i < MILLION; i++) {
    stack.push(i);
  }
  for (let i = 0; i < MILLION; i++) {
    stack.pop();
  }
  this.val = stack;
});

suite.add('Native JS Array 1M push & pop', function() {
  const arr = [];
  for (let i = 0; i < MILLION; i++) {
    arr.push(i);
  }
  for (let i = 0; i < MILLION; i++) {
    arr.pop();
  }
  this.val = arr;
});

export { suite };